flyteplugins/go/tasks/plugins/k8s/kfoperators/common/common_operator.go

```diff
@@ -254,27 +251,101 @@ func ParseRestartPolicy(flyteRestartPolicy kfplugins.RestartPolicy) commonOp.Res
 }
 
 // OverrideContainerSpec overrides the specified container's properties in the given podSpec. The function
-// updates the image, resources and command arguments of the container that matches the given containerName.
-func OverrideContainerSpec(podSpec *v1.PodSpec, containerName string, image string, resources *core.Resources, args []string) error {
+// updates the image and command arguments of the container that matches the given containerName.
+func OverrideContainerSpec(podSpec *v1.PodSpec, containerName string, image string, args []string) error {
 	for idx, c := range podSpec.Containers {
 		if c.Name == containerName {
 			if image != "" {
 				podSpec.Containers[idx].Image = image
 			}
-			if resources != nil {
-				// if resources requests and limits both not set, we will not override the resources
-				if len(resources.Requests) >= 1 || len(resources.Limits) >= 1 {
-					resources, err := flytek8s.ToK8sResourceRequirements(resources)
-					if err != nil {
-						return flyteerr.Errorf(flyteerr.BadTaskSpecification, "invalid TaskSpecification on Resources [%v], Err: [%v]", resources, err.Error())
-					}
-					podSpec.Containers[idx].Resources = *resources
-				}
-			}
 			if len(args) != 0 {
 				podSpec.Containers[idx].Args = args
 			}
 		}
 	}
 	return nil
 }
+
+func ToReplicaSpec(ctx context.Context, taskCtx pluginsCore.TaskExecutionContext, primaryContainerName string) (*commonOp.ReplicaSpec, error) {
+	podSpec, objectMeta, oldPrimaryContainerName, err := flytek8s.ToK8sPodSpec(ctx, taskCtx)
+	if err != nil {
+		return nil, flyteerr.Errorf(flyteerr.BadTaskSpecification, "Unable to create pod spec: [%v]", err.Error())
+	}
+
+	OverridePrimaryContainerName(podSpec, oldPrimaryContainerName, primaryContainerName)
+
+	cfg := config.GetK8sPluginConfig()
+	objectMeta.Annotations = utils.UnionMaps(cfg.DefaultAnnotations, objectMeta.Annotations, utils.CopyMap(taskCtx.TaskExecutionMetadata().GetAnnotations()))
+	objectMeta.Labels = utils.UnionMaps(cfg.DefaultLabels, objectMeta.Labels, utils.CopyMap(taskCtx.TaskExecutionMetadata().GetLabels()))
+
+	replicas := int32(0)
+	return &commonOp.ReplicaSpec{
+		Replicas: &replicas,
+		Template: v1.PodTemplateSpec{
+			ObjectMeta: *objectMeta,
+			Spec:       *podSpec,
+		},
+		RestartPolicy: commonOp.RestartPolicyNever,
+	}, nil
+}
+
+type kfDistributedReplicaSpec interface {
+	GetReplicas() int32
+	GetImage() string
+	GetResources() *core.Resources
+	GetRestartPolicy() kfplugins.RestartPolicy
+}
+
+type allowsCommandOverride interface {
+	GetCommand() []string
+}
+
+func ToReplicaSpecWithOverrides(ctx context.Context, taskCtx pluginsCore.TaskExecutionContext, rs kfDistributedReplicaSpec, primaryContainerName string, isMaster bool) (*commonOp.ReplicaSpec, error) {
+	taskCtxOptions := []flytek8s.PluginTaskExecutionContextOption{}
+	// Master should always run as non-interruptible
+	if isMaster {
+		taskCtxOptions = append(taskCtxOptions, flytek8s.WithInterruptible(false))
```

Comment:
Note that "master" pods will be tagged as non-interruptible.

Reply:
I think for PyTorch this doesn't make sense as the entire pytorch job needs to be restarted when one of the workers is preempted. This, however, is not done by the master replica but the kubeflow training operator.
I think in general for pytorch jobs the differentiation between master and worker doesn't make sense. The fact that pytorch jobs with an elastic config don't have any master replicas at all anymore bespeaks that.